### ISSUE TITLE
Broken/Unsupported Prediction on entities?

### ISSUE THREAD

**Issue by BoZ-Niko663** [Facepunch/garrysmod-issues#4544]
### Details
Happening on the latest non-dev Garry's Mod May 2020 update.

Apologies in advance for this wall of text, but I kinda need an answer.

So I am updating my Halo 2 SWEPS as I've noticed they are not properly networked and often succumb to lag in Multiplayer. I was using a local server and a Player 2 via the handy "-multirun" command to test anything that would need to be tested. It was all going well until I noticed something with the needler. The needler if you are familiar is a weapon that fires projectiles that damage an entity and after a certain amount of needles get fired into an entity it explodes in something known as a "Supercombine". So I was testing base damage, entities should take 7 damage per non-supercombine shot. Shooting at Alyx due to her fast health regeneration, I had the hit numbers mod to let me know how much damage an entity takes, I noticed Alyx spewed two "-7" indiccators when I shot her, then after a few more shots she spewed 3 "-7" indicators at a time! At one point there was probably a mini lag spike or something because I fired the needler at a combine with 100 health and he died instantly with like 12 "-7" indicators spewing out of him. Obviously, for balacing and aesthetic reasons, this is not OK, so I decided to use Prediction, it didnt work, still the same problem. I tried many different variations including is SERVER then, if IsFirstTimePredicted, if !CLIENT then ETC. But none of them nor any variation seemed to make a difference so I enabled "cl_predictionlist" and found that no matter how many needler rounds I fired, none of them would appear on the list, I tried other addons and those entities did not appear on the list either! So it seems Prediction is not supported for these entities, but those entities don't have any lag/dual damage problems. I tried the m9k specialties harpoon and didnt get a single dual damage problem. So am I just dumb and dont know how to use Prediction on entities or does it just not exist? If it doesnt, it really should.

TLDR ; My needler projectile entity occasionally deals damage multiple times in Multiplayer, like a Prediction issue, nothing I can do fixes it.

### Steps to reproduce
create    "lua/entities/needle_h2" create a "shared.lua" file and paste this code

____________________

ENT.Type 			= "anim"
ENT.Base 			= "base_gmodentity"
ENT.PrintName			= "Needler Bolt from Halo 2"
ENT.Author			= "[BoZ]Niko663"

ENT.Spawnable			= false
ENT.AdminSpawnable		= false

if (CLIENT) then
killicon.Add( "needle_h2", "VGUI/hud/halo2_needler", color_white )
end

AddCSLuaFile( "shared.lua" )
include( 'shared.lua' )

function ENT:SpawnFunction( ply, tr )

	if ( !tr.Hit ) then return end
	
	local SpawnPos = tr.HitPos + tr.HitNormal * 16
	
	local ent = ents.Create( "needle_h2" )
		ent:SetPos( SpawnPos )
	ent:Spawn()
	ent:Activate()
	
	return ent
	
end

/*---------------------------------------------------------
   Name: Initialize
---------------------------------------------------------*/
function ENT:Initialize()
		local owner = self.Entity.Owner
	self.Entity:SetModel( "models/items/crossbowrounds.mdl" )
	self.Entity:SetSolid(SOLID_VPHYSICS)
	self.Entity:PhysicsInit(MOVETYPE_VPHYSICS)
	self.Entity:SetMoveType(MOVETYPE_VPHYSICS)

	// Wake the physics object up.
	local phys = self.Entity:GetPhysicsObject()
	if (phys:IsValid()) then
		phys:Wake()
		phys:EnableGravity(false)
		phys:SetMass(1)

Glow = ents.Create("env_sprite")
Glow:SetKeyValue("model","orangecore2.vmt")
Glow:SetKeyValue("rendercolor","255 0 191")
Glow:SetKeyValue("scale","0.6")
Glow:SetPos(self.Entity:GetPos())
Glow:SetParent(self.Entity)
Glow:Spawn()
Glow:Activate()

		self:EmitSound("Halo2_Needler.FlyBy")
end
	self.Angles = self.Entity:GetAngles()
	self.Entity:Fire("kill","",10)
end

inactive = false
/*---------------------------------------------------------
   Name: PhysicsCollide
---------------------------------------------------------*/
function ENT:PhysicsCollide( data, physobj )

local target = data.HitEntity

	if target:IsNPC() and (SERVER) and target:Health() > 40 and IsValid(target) then
	if IsFirstTimePredicted() then
	target:TakeDamage(7,self.Entity.Owner,self.Entity.Owner)
	self:EmitSound("halo2/needler_impact.wav")
	self:StopSound("Halo2_Needler.FlyBy")
	self.Entity:Fire("kill","",0)
	end
elseif target:IsNPC() and (SERVER) and target:Health() <= 40 and IsValid(target) then
	if IsFirstTimePredicted() then
    self:SuperCombineH2()
	self:StopSound("Halo2_Needler.FlyBy")
	end
elseif target:IsPlayer() and target:Health() > 50 and IsValid(target) then
    if IsFirstTimePredicted() then
	target:TakeDamage(8,self.Entity.Owner,self.Entity.Owner)
	self:EmitSound("halo2/needler_impact.wav")
	self:StopSound("Halo2_Needler.FlyBy")
	self.Entity:Fire("kill","",0)
	end
elseif target:IsPlayer() and target:Health() <= 50 and IsValid(target) then
	if IsFirstTimePredicted() then
    self:SuperCombineH2()
	self:StopSound("Halo2_Needler.FlyBy")
	end
else
	self:EmitSound("halo2/needler_impact.wav")
	self:StopSound("Halo2_Needler.FlyBy")
	local needle = ents.Create( "needle_inactive_h2" )
	needle:SetPos( self:GetPos() )
	needle:SetAngles( self:GetAngles() )
	needle:Spawn()
	needle:Activate()
	self.Entity:Fire("kill","",0)
end
end

function ENT:SuperCombineH2()

		local effectdata = EffectData()
					effectdata:SetOrigin(self:GetPos())			// Where is hits
					effectdata:SetNormal(Vector(0,0,1))		// Direction of particles
					effectdata:SetEntity(self.Entity)		// Who done it?
					effectdata:SetScale(1.4)			// Size of explosion		// Length of explosion trails
					util.Effect( "supercombineh2", effectdata )

	self:EmitSound("Halo2_Needler.SuperCombine")
	util.BlastDamage(self.Entity, self.Owner, self.Entity:GetPos(), 160, 160)
	util.ScreenShake(self.Entity:GetPos(), 250, 250, 1.25, 250)
	self.Entity:Fire("kill","",0)
end

/*---------------------------------------------------------
   Name: OnTakeDamage
---------------------------------------------------------*/
function ENT:OnTakeDamage( dmginfo )
end

/*---------------------------------------------------------
   Name: Use
---------------------------------------------------------*/
function ENT:Use( activator, caller )
end  

function ENT:OnRemove()
self:StopSound("Halo2_Needler.FlyBy")
end
_________________________________

Create a SWEP with automatic fire and the fire rate of a counter strike m249 or around there. Call the "FireNeedle" function in PrimaryAttack

function SWEP:FireNeedle()
	local aim = self.Owner:GetAimVector()
	local side = aim:Cross(Vector(0,0,1))
	local up = side:Cross(aim)
	local pos = self.Owner:GetShootPos() + side * 10 + up * -6

	if SERVER then
	local rocket = ents.Create(self.Secondary.Round)
	if !rocket:IsValid() then return false end
	rocket:SetAngles(self.Owner:GetAimVector():Angle(90,90,0))
	rocket:SetPos(pos)
	rocket:SetOwner(self.Owner)
	rocket:Spawn()
	rocket.Owner = self.Owner
	rocket:Activate()
	eyes = self.Owner:EyeAngles()
		local phys = rocket:GetPhysicsObject()
			phys:SetVelocity(self.Owner:GetAimVector() * 1100)
	end
		if SERVER and !self.Owner:IsNPC() then
		local anglo = Angle(0, 0, 0)		
		self.Owner:ViewPunch(anglo)
		end

end
___________________________

Once these are set up, start a Local or Peer to Peer server, doesnt matter which.
Equip the SWEP and fire it at NPCs or bots, even with a good connection, every once in a while you will notice the entity you shoot at take damage multiple times!

**Comment by robotboy655** [comment]
This is not a place to get modding help. You can try asking the community for help with modding on our official Discord server:
https://discord.gg/gmod

I'd say likely the issue is coming from `PhysicsCollide` being called multiple times, which should be an easy fix for you if you know what you are doing.
